src/main/java/org/broadinstitute/hellbender/utils/hmm/segmentation/HiddenMarkovModelPostProcessor.java

```diff
@@ -0,0 +1,1145 @@
+     * @param allSegments map with all segment lists coming from different samples.
+     * @return never null.
+     */
+    private Iterator<HiddenStateSegmentRecord<S, T>> composeTargetSortedSegmentRecordIterator(
+            final TargetCollection<T> masterTargetCollection,
+            final Map<String, List<HiddenStateSegment<S, T>>> allSegments) {
+        final List<Iterator<HiddenStateSegmentRecord<S, T>>> allSegmentRecordIterators =
+                allSegments.entrySet().stream()
+                        .map(e -> e.getValue().stream()
+                                .map(s -> new HiddenStateSegmentRecord<>(e.getKey(), s))
+                                .iterator())
+                        .collect(Collectors.toList());
+
+        final Comparator<HiddenStateSegmentRecord<S, T>> recordComparator = (o1, o2) -> {
+            final HiddenStateSegment<S, T> s1 = o1.getSegment();
+            final HiddenStateSegment<S, T> s2 = o2.getSegment();
+            // Using the index-range.from we make sure we sort first by the contig over
+            // the start position in the same order as contigs are present in
+            // the input data or target collection.
+            final IndexRange ir1 = masterTargetCollection.indexRange(o1.getSegment());
+            final IndexRange ir2 = masterTargetCollection.indexRange(o2.getSegment());
+
+            final int fromCmp = Integer.compare(ir1.from, ir2.from);
+            if (fromCmp != 0) {
+                return fromCmp;
+            }
+            // if fromCmp == 0, they must be in the same contig,
+            // then we can sort based on segment start and then end.
+            final int startCmp = Integer.compare(s1.getStart(), s2.getStart());
+            if (startCmp != 0) {
+                return startCmp;
+            }
+            return Integer.compare(s1.getEnd(), s2.getEnd());
+        };
+
+        return Iterators.mergeSorted(allSegmentRecordIterators, recordComparator);
+    }
+
+    /**
+     * Writes sorted segments into the output file provided by the user.
+     *
+     * @param allSegmentsSorted iterator that must produce segment records in order.
+     * @param writer a table writer
+     * @throws org.broadinstitute.hellbender.exceptions.UserException.CouldNotCreateOutputFile if there is any
+     * issue creating or writing into the output file provided by the user.
+     */
+    private void writeSegments(final Iterator<HiddenStateSegmentRecord<S, T>> allSegmentsSorted,
+                               final HiddenStateSegmentRecordWriter<S, T> writer) {
+        try {
+            long count = 0;
+            final Set<String> samples = new HashSet<>();
+            while (allSegmentsSorted.hasNext()) {
+                final HiddenStateSegmentRecord<S, T> record = allSegmentsSorted.next();
+                samples.add(record.getSampleName());
+                writer.writeRecord(record);
+                count++;
```

Comment:
No, it is the number of elements in the iterator (count = number of segments); so we must keep track of it explicitly as we iterate.